a11y-keyboard: Add support for feedback when a key is pressed and CapsLock is active

cwendling opened the pull request:
Adds support for beeping when a key is pressed while CapsLock is active.

This is useful for people with limited to no sight that might not remember or noticed CapsLock is active, and can prevent them from typing longish pieces of text before noticing it didn't type as expected.

This depends on https://github.com/mate-desktop/mate-desktop/pull/450.  See also https://github.com/mate-desktop/mate-control-center/pull/583.

cwendling commented on plugins/a11y-keyboard/msd-a11y-keyboard-atspi.c:
```diff
@@ -0,0 +1,190 @@
+#include <gdk/gdk.h>
+#include <atspi/atspi.h>
+
+/* See https://gitlab.gnome.org/GNOME/at-spi2-core/-/issues/22 */
+#define DESTROYING_ATSPI_LISTENER_CRASHES \
+        (! (defined (HAVE_LIBATSPI_NOBUG22) && HAVE_LIBATSPI_NOBUG22))
```
This is sad, but unfortunately there's no released version with a fix, so we'll have to live with the workaround.

cwendling commented:
Travis fails because of the mate-desktop dependency that isn't satisfied there.

raveit65 commented:
When mate-desktop PR is merged and mate-desktop-1.25.0 is released, we need to update `build.yml` of mate-settings-deamon.
See https://github.com/mate-desktop/mate-settings-daemon/blob/master/.build.yml#L145
```
- cd ${START_DIR}
  - '[ -f mate-desktop-1.23.2.tar.xz ] || curl -Ls -o mate-desktop-1.23.2.tar.xz http://pub.mate-desktop.org/releases/1.23/mate-desktop-1.23.2.tar.xz'
  - tar xf mate-desktop-1.23.2.tar.xz
  - cd mate-desktop-1.23.2
  - if [ ${DISTRO_NAME} == "debian" -o ${DISTRO_NAME} == "ubuntu" ];then
  -     ./configure --prefix=/usr --libdir=/usr/lib/x86_64-linux-gnu --libexecdir=/usr/lib/x86_64-linux-gnu
  - else
  -     ./configure --prefix=/usr
  - fi
  - make
  - make install
```
This needs to be updated to mate-desktop-1.25.0.tar.xz

raveit65 commented:
@cwendling 
mate-desktop-1.25 is released.

cwendling commented:
@raveit65 thanks, apparently it works now :)

sc0w commented:
new warnings in the logs:

```
msd-a11y-keyboard-atspi.c:84:5: warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if DESTROYING_ATSPI_LISTENER_CRASHES
    ^
msd-a11y-keyboard-atspi.c:32:13: note: expanded from macro 'DESTROYING_ATSPI_LISTENER_CRASHES'
        (! (defined (HAVE_LIBATSPI_NOBUG22) && HAVE_LIBATSPI_NOBUG22))
            ^

msd-a11y-keyboard-atspi.c:133:7: warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if ! DESTROYING_ATSPI_LISTENER_CRASHES
      ^
msd-a11y-keyboard-atspi.c:32:13: note: expanded from macro 'DESTROYING_ATSPI_LISTENER_CRASHES'
        (! (defined (HAVE_LIBATSPI_NOBUG22) && HAVE_LIBATSPI_NOBUG22))
            ^

msd-a11y-keyboard-atspi.c:152:5: warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
#if DESTROYING_ATSPI_LISTENER_CRASHES
    ^
msd-a11y-keyboard-atspi.c:32:13: note: expanded from macro 'DESTROYING_ATSPI_LISTENER_CRASHES'
        (! (defined (HAVE_LIBATSPI_NOBUG22) && HAVE_LIBATSPI_NOBUG22))
            ^

msd-a11y-keyboard-manager.c:992:49: warning: unused parameter 'key' [-Wunused-parameter]
                        gchar                  *key,
                                                ^
```

sc0w commented:
NOTE: we are seeing now warnings in the logs unrelated to mate, warnings in the library atspi, for example:

```
/usr/include/at-spi-2.0/atspi/atspi-collection.h:43:1: warning: function declaration isn't a prototype [-Wstrict-prototypes]
   43 | GType atspi_collection_get_type ();
      | ^~~~~
```

@mate-desktop/core-team do you know how to silence the warnings of libraries at travis?

rbuj commented:
I don’t think it’s necessary to do this, as the warnings are sometimes due to other parts of code, such as the expansion of CLAMP macro and unsigned integers.

cwendling commented:
> new warnings in the logs:
> 
> ```
> msd-a11y-keyboard-atspi.c:84:5: warning: macro expansion producing 'defined' has undefined behavior [-Wexpansion-to-defined]
> […]
> ```

Oopsie, that's what I get for trying to get this extra thing in a little too fast (as the upstream bug wasn't fixed when I wrote the bulk of the patch, I didn't have a check for it set up apart the define hard-coded)… anyway, should be fixed now, I switched to a plain conditional definition.

> ```
> msd-a11y-keyboard-manager.c:992:49: warning: unused parameter 'key' [-Wunused-parameter]
> ```

Indeed.  I fixed this by marking the parameter as unused, as it's perfectly legitimate as it's a callback and this parameter is not useful in this case.

> @mate-desktop/core-team do you know how to silence the warnings of libraries at travis?

AFAIK you can't, short of getting the headers of those libraries warnings-free.  I tried to deal with this a while back, and I couldn't find a proper solution.  The only remotely close one I encountered was trying to get the compiler recognize the header as "system" ones, but that's actually a lot different than just "installed" headers vs local ones, and IIRC had practical drawbacks.

rbuj commented on plugins/a11y-keyboard/msd-a11y-keyboard-atspi.c:
```diff
@@ -0,0 +1,193 @@
+
+#include <glib-object.h>
+
+#if HAVE_LIBATSPI
```
It should be `#ifdef HAVE_LIBATSPI`

cwendling replied:
Indeed, fixed

rbuj commented on plugins/a11y-keyboard/msd-a11y-keyboard-atspi.c:
```diff
@@ -0,0 +1,193 @@
+#include <atspi/atspi.h>
+
+/* See https://gitlab.gnome.org/GNOME/at-spi2-core/-/issues/22 */
+#if defined (HAVE_LIBATSPI_NOBUG22) && HAVE_LIBATSPI_NOBUG22
+#       define DESTROYING_ATSPI_LISTENER_CRASHES 0
+#else
+#       define DESTROYING_ATSPI_LISTENER_CRASHES 1
+#endif
```
You can use `#ifdef HAVE_LIBATSPI_NOBUG22` and `#ifndef HAVE_LIBATSPI_NOBUG22` since you have defined it on configure.ac

cwendling replied:
Yeah indeed, now it's checked in the configure script that would make sense.  I did that, and changed the config.h name to be clearer in the code

rbuj commented on plugins/a11y-keyboard/msd-a11y-keyboard-atspi.c:
```diff
@@ -0,0 +1,193 @@
+
+static gboolean
+on_key_press_event (const AtspiDeviceEvent *event,
+                    void                   *user_data G_GNUC_UNUSED)
```
IMHO, it's not a good idea to silence the warnings.

sc0w replied:
LGTM this change

cwendling replied:
@rbuj why?  IMO `-Wunused-parameter` is a fair warning that can help spot bugs and suggest refactoring, but here as it's a callback it has a specific signature I can't alter, yet I really don't need it here, as I know it's value (it's useful when the signal can be triggered for several GSettings keys, but here it's set up to only be emitted for a specific one, so it's pointless to check again).
So IMO it's a good idea to silence this instance of the warning as it's a kind of false positive, and would make the other warnings harder to spot in the mix.

rbuj replied:
If you want to filter out this warning type, you can use "-Wno-unused-variable", if you don't need a parameter you can use the function cast macro G_CALLBACK or use void for overriding the parameter within the function body, but using G_GNUC_UNUSED silents this warnings, which cannot be detected if you add G_GNUC_UNUSED.

rbuj replied:
G_GNUC_UNUSED only appears 27 times:
```
$ grep -R G_GNUC_UNUSED --include=*.c
mate-session-manager/mate-session/test-inhibit.c:        const char *startup_id G_GNUC_UNUSED;
marco/src/ui/tabpopup.c:                             GdkScreen *old_screen G_GNUC_UNUSED,
marco/src/ui/tabpopup.c:                             gpointer   data G_GNUC_UNUSED)
marco/src/core/main.c:      G_GNUC_UNUSED int dummy;
libmatemixer/backends/oss/oss-backend.c:                                                  guint             index) G_GNUC_UNUSED;
mate-terminal/src/terminal-screen-container.c:                                      GtkPolicyType hpolicy G_GNUC_UNUSED,
engrampa/src/fr-init.c:G_GNUC_UNUSED static void
engrampa/src/fr-init.c:G_GNUC_UNUSED static gboolean
engrampa/src/eggfileformatchooser.c:                             gpointer value G_GNUC_UNUSED,
engrampa/src/eggfileformatchooser.c:                GtkTreePath  *path G_GNUC_UNUSED,
engrampa/src/eggfileformatchooser.c:                   GtkTreePath  *path G_GNUC_UNUSED,
engrampa/src/eggfileformatchooser.c:                                            guint                 parent G_GNUC_UNUSED,
caja/src/caja-application.c:                  GParamSpec     *unused G_GNUC_UNUSED,
pluma/pluma/pluma-print-preview.c:dummy_write_func (G_GNUC_UNUSED gpointer      closure,
pluma/pluma/pluma-print-preview.c:		  G_GNUC_UNUSED const guchar *data,
pluma/pluma/pluma-print-preview.c:		  G_GNUC_UNUSED guint         length)
mate-panel/applets/notification_area/testtray.c:signal_handler (gpointer data G_GNUC_UNUSED)
mate-panel/mate-panel/panel-test-applets.c:G_GNUC_UNUSED void on_execute_button_clicked (GtkButton *button, gpointer dummy);
mate-panel/mate-panel/panel-test-applets.c:G_GNUC_UNUSED void
mate-settings-daemon/plugins/mouse/msd-mouse-manager.c:        G_GNUC_UNUSED XEventClass class_presence;
mate-settings-daemon/plugins/mouse/msd-mouse-manager.c:        G_GNUC_UNUSED  int xi_presence;
mate-settings-daemon/plugins/xsettings/fontconfig-monitor.c:stuff_changed (GFileMonitor *monitor G_GNUC_UNUSED,
mate-settings-daemon/plugins/xsettings/fontconfig-monitor.c:               GFile *file G_GNUC_UNUSED,
mate-settings-daemon/plugins/xsettings/fontconfig-monitor.c:               GFile *other_file G_GNUC_UNUSED,
mate-settings-daemon/plugins/xsettings/fontconfig-monitor.c:               GFileMonitorEvent event_type G_GNUC_UNUSED,
mate-settings-daemon/plugins/a11y-keyboard/msd-a11y-keyboard-manager.c:        G_GNUC_UNUSED XEventClass class_presence;
mate-settings-daemon/plugins/a11y-keyboard/msd-a11y-keyboard-manager.c:        G_GNUC_UNUSED int xi_presence;
```

cwendling replied:
> If you want to filter out this warning type, you can use "-Wno-unused-parameter",

Yes, but that will filter useful warnings as well.  The point of marking the ones that are legitimate is to make it clear to everyone it's known and OK, and not to clutter the compiler output.  It's basically impossible to spot 1 useful warning if you have a thousand false positive, so the useful one would just get lost.

> if you don't need a parameter you can use the function cast macro G_CALLBACK

Yes, in this case it would work.  It's slightly wrong on paper as it still is calling an incorrectly typed function, but in practice it'll work as GTK itself uses this kind of trick so all architectures it runs on ought to be OK with this.

> or use void for overriding the parameter within the function body, but using G_GNUC_UNUSED silents these warnings, which cannot be detected if you add G_GNUC_UNUSED.

I don't get this. What's the difference between making dummy use of the argument or marking it G_GNUC_UNUSED?  It even seems easier to me to spot G_GNUC_UNUSED than `(void) param;`, so I'd expect it to be *easier* to detect (whatever you mean by that).

> G_GNUC_UNUSED only appears 27 times in about 665000 lines of code:
>
> without including the 27 muted warnings there are 6776 warnings about -Wunused-parameter in our current build logs

What point are you trying to make?  If it's that G_GNUC_UNUSED is not used much in the codebase, OK, but that doesn't necessarily mean it *should* not be.  And if you tell me there are 6776 warnings about this, I wonder if we indeed shouldn't start doing something about this (not saying it's a priority, but still).
I don't see your data as suggesting we should not be using any mean of silencing unused parameter warnings, rather the contrary.  Which technique to use is another question, and if there's a consensus on that I'm happy to follow it, but for the moment I still like G_GNUC_UNUSED.

sc0w replied:
I agree 100% with @cwendling 

@cwendling maybe you like to review https://github.com/mate-desktop/mate-utils/pull/300 which can be handled easily with G_GNUC_UNUSED :)

rbuj replied:
What I’m pointing you is that the codebase prefers to show all warnings instead of cheating and hiding the ones you don't like to show.

Once we have reached this point in the conversation, I think the best is to seek the opinion of other team members. We have both put forward enough arguments for and against the use of the `G_GNUC_UNUSED` macro which is expanded to `unused`.
- https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-unused-function-attribute
- https://clang.llvm.org/docs/AttributeReference.html#maybe-unused-unused

I will respect the decision made by the team.

@mate-desktop/core-team Can you help us decide what is the best option in this case?

sc0w replied:
@rbuj 

> What I’m pointing you is that the codebase prefers to show all warnings instead of cheating and hiding the ones you don't like to show.

it is false positive warning, I don't like to see false positive warnings

at my point of view, the author of PR can decide

> I will respect the decision made by the team.

by the moment, 2 members versus 1

lukefromdc replied:
I don't know enough about this with the warnings to hold a useful opinion

rbuj commented on plugins/a11y-keyboard/msd-a11y-keyboard-atspi.c:
```diff
@@ -0,0 +1,193 @@
+G_DEFINE_TYPE (MsdA11yKeyboardAtspi, msd_a11y_keyboard_atspi, G_TYPE_OBJECT)
+
+static void
+msd_a11y_keyboard_atspi_class_init (MsdA11yKeyboardAtspiClass *klass G_GNUC_UNUSED)
+{
+}
+
+static void
+msd_a11y_keyboard_atspi_init (MsdA11yKeyboardAtspi *self G_GNUC_UNUSED)
+{
+}
+
+void
+msd_a11y_keyboard_atspi_start (MsdA11yKeyboardAtspi *self G_GNUC_UNUSED)
+{
+}
+
+void
+msd_a11y_keyboard_atspi_stop (MsdA11yKeyboardAtspi *self G_GNUC_UNUSED)
+{
+}
```
IMHO, it's not a good idea to silence the warnings.

rbuj replied:
See https://github.com/mate-desktop/mate-settings-daemon/pull/330/files#r467228696

cwendling replied:
This is no longer a concern as this code has been removed.

rbuj commented on plugins/a11y-keyboard/msd-a11y-keyboard-manager.c:
```diff
@@ -984,6 +987,17 @@ keyboard_callback (GSettings              *settings,
 
+static void
+capslock_beep_callback (GSettings              *settings,
+                        gchar                  *key G_GNUC_UNUSED,
```
IMHO, it's not a good idea to silence the warnings.

rbuj commented on plugins/a11y-keyboard/msd-a11y-keyboard-atspi.h:
```diff
@@ -0,0 +1,55 @@
+
+G_BEGIN_DECLS
+
+#define MSD_TYPE_A11Y_KEYBOARD_ATSPI            (msd_a11y_keyboard_atspi_get_type ())
+#define MSD_A11Y_KEYBOARD_ATSPI(o)              (G_TYPE_CHECK_INSTANCE_CAST ((o), MSD_TYPE_A11Y_KEYBOARD_ATSPI, MsdA11yKeyboardAtspi))
+#define MSD_A11Y_KEYBOARD_ATSPI_CLASS(k)        (G_TYPE_CHECK_CLASS_CAST((k), MSD_TYPE_A11Y_KEYBOARD_ATSPI, MsdA11yKeyboardAtspiClass))
+#define MSD_IS_A11Y_KEYBOARD_ATSPI(o)           (G_TYPE_CHECK_INSTANCE_TYPE ((o), MSD_TYPE_A11Y_KEYBOARD_ATSPI))
+#define MSD_IS_A11Y_KEYBOARD_ATSPI_CLASS(k)     (G_TYPE_CHECK_CLASS_TYPE ((k), MSD_TYPE_A11Y_KEYBOARD_ATSPI))
+#define MSD_A11Y_KEYBOARD_ATSPI_GET_CLASS(o)    (G_TYPE_INSTANCE_GET_CLASS ((o), MSD_TYPE_A11Y_KEYBOARD_ATSPI, MsdA11yKeyboardAtspiClass))
+
+typedef struct MsdA11yKeyboardAtspiPrivate MsdA11yKeyboardAtspiPrivate;
+
+typedef struct
+{
+        GObject                         parent;
+        MsdA11yKeyboardAtspiPrivate    *priv;
+} MsdA11yKeyboardAtspi;
```
You can use boilerplate.

cwendling replied:
Good idea, done.  I tend to forget these handy macro were added to GObject as I started using it way before that, so got used to writing the boilerplate by hand…

rbuj commented on plugins/a11y-keyboard/Makefile.am:
```diff
@@ -42,6 +42,8 @@ liba11y_keyboard_la_SOURCES = 		\
 	msd-a11y-keyboard-plugin.c	\
 	msd-a11y-keyboard-manager.h	\
 	msd-a11y-keyboard-manager.c	\
+	msd-a11y-keyboard-atspi.h	\
+	msd-a11y-keyboard-atspi.c	\
```
These files should not be included based on HAVE_LIBATSPI def, you must also remove the empty functions on msd-a11y-keyboard-atspi.c if the library is not available, as well as surrounding their usage in plugins/a11y-keyboard/msd-a11y-keyboard-manager.c and plugins/a11y-keyboard/msd-a11y-preferences-dialog.c, what's more you have to hide the setting in the preference dialog if the library is not available.

cwendling replied:
The point in having them unconditionally was to make it easier on the manager.  I can indeed not build empty functions when not available if you prefer and add guards on the other end.

rbuj replied:
FMPOV, it's better, otherwise you're making [software bloat](https://en.wikipedia.org/wiki/Software_bloat)

cwendling replied:
I don't completely agree with software bloat here, although it's technically correct: there's another factor to consider, which is maintainability.  Code riddled with `#ifdef`s can quickly become a nightmare to navigate and understand, so sometimes it's better to keep the conditionals at a lower more abstract level.

Anyway, in this case I don't think it hurts readability and maintenance too much, so I did it.

So both concerns in your comment should be addressed now.

rbuj commented on configure.ac:
```diff
@@ -115,6 +116,18 @@ fi
+dnl - Check for libatspi
+dnl ---------------------------------------------------------------------------
+
+PKG_CHECK_MODULES(LIBATSPI, atspi-2 >= $LIBATSPI_REQUIRED_VERSION,
+          [AC_DEFINE(HAVE_LIBATSPI, 1, [Define if libatspi is available])
+          have_libatspi=yes], have_libatspi=no)
+AM_CONDITIONAL([HAVE_LIBATSPI], [test "x$have_libatspi" = xyes])
+PKG_CHECK_EXISTS([atspi-2 > 2.36.0],
+                 [AC_DEFINE([DESTROYING_ATSPI_LISTENER_DOES_NOT_CRASH], [1], [Define if libatspi does not have bug 22])],
+                 [])
+
```
The user must be able to enable or disable this feature, then check if the library is available and show if this is enabled in the configuration summary.

cwendling replied:
OK, will do

cwendling replied:
done

rbuj commented on plugins/a11y-keyboard/Makefile.am:
```diff
@@ -55,6 +61,7 @@ liba11y_keyboard_la_CPPFLAGS = \
 liba11y_keyboard_la_CFLAGS =		\
 	$(SETTINGS_PLUGIN_CFLAGS)	\
 	$(LIBNOTIFY_CFLAGS)		\
+	$(LIBATSPI_CFLAGS)		\
```
The use of this variable is conditional

cwendling replied:
It makes sense for semantics, but FWIW if HAVE_LIBATSPI is not set it won't be either and thus will expand to an empty value.

Will fix anyway.

rbuj commented on plugins/a11y-keyboard/Makefile.am:
```diff
@@ -66,6 +73,7 @@ liba11y_keyboard_la_LDFLAGS = 		\
 liba11y_keyboard_la_LIBADD  = 		\
 	$(SETTINGS_PLUGIN_LIBS)		\
 	$(LIBNOTIFY_LIBS)		\
+	$(LIBATSPI_LIBS)		\
```
The use of this variable is conditional

rbuj commented on plugins/a11y-keyboard/msd-a11y-keyboard-atspi.c:
```diff
@@ -0,0 +1,158 @@
+#include <glib-object.h>
+#include <gdk/gdk.h>
+#include <atspi/atspi.h>
+
+
```
IMHO, you don't need to use two consecutive blank lines.

cwendling replied:
nitpicky :)  But yeah we don't have those in other similar files, so I dropped one

rbuj commented on plugins/a11y-keyboard/msd-a11y-keyboard-atspi.c:
```diff
@@ -0,0 +1,158 @@
+        /* register listeners for all keys with CAPS_LOCK modifier */
+        for (AtspiKeyMaskType mod_mask = 0; mod_mask < 256; mod_mask++)
+        {
+                if (! (mod_mask & (1 << ATSPI_MODIFIER_SHIFTLOCK)))
```
`(1 << ATSPI_MODIFIER_SHIFTLOCK)` == `SPI_KEYMASK_SHIFTLOCK`?
```
registryd/keymasks.h:#define SPI_KEYMASK_SHIFTLOCK (1<<1)  /* LockMask */
```
If they are the same I prefer `SPI_KEYMASK_SHIFTLOCK`

cwendling replied:
It's not part of the API, is it?

rbuj replied:
No, it's not available in the [API](https://developer.gnome.org/libatspi/stable/libatspi-atspi-constants.html), therefore it's better to use `1 << ATSPI_MODIFIER_SHIFTLOCK` as mask maker, instead of adding `#define SPI_KEYMASK_SHIFTLOCK (1<<1)` a the top of the file, as a maker for (de)registering key event listeners.
```
#define ATSPI_MODIFIERTYPE_COUNT (7+1)
```
2^8=256, possible masks and you only (de)register those that have the bit ATSPI_MODIFIER_SHIFTLOCK set to 1, ie 128 event listeners.

I don’t think it’s the most efficient way to deal with a modifier key mask but it seems valid at first glance.

cwendling replied:
Yeah I know this API seems a bit weird, but it seems to be how it's supposed to be used -- and how other clients like Orca deal with it.

rbuj commented on plugins/a11y-keyboard/msd-a11y-keyboard-manager.c:
```diff
@@ -49,9 +49,13 @@
 #include "msd-a11y-preferences-dialog.h"
 
 #define CONFIG_SCHEMA "org.mate.accessibility-keyboard"
+#define KEY_CAPSLOCK_BEEP_ENABLED "capslock-beep-enable"
```
You always use this macro within blocks where HAVE_LIBATSPI  is checked, so to be consistent you would have to declare it based on HAVE_LIBATSPI

cwendling replied:
I think it starts getting a little too many preprocessor conditionals, but indeed it makes sense.

BTW, not that it matters for this one, but there's the same issue with `NOTIFICATION_TIMEOUT` on the next line.

rbuj commented on plugins/a11y-keyboard/msd-a11y-preferences-dialog.c:
```diff
@@ -49,6 +49,7 @@
 #define KEY_BOUNCE_KEYS_ENABLED      "bouncekeys-enable"
 #define KEY_SLOW_KEYS_ENABLED        "slowkeys-enable"
 #define KEY_MOUSE_KEYS_ENABLED       "mousekeys-enable"
+#define KEY_CAPSLOCK_BEEP_ENABLED    "capslock-beep-enable"
```
This macro has been defined in another file.

cwendling replied:
Yes, but I'm quite unsure where we should share those.  Actually we could share quite a few of them instead of hard-coding several strings, but that's IMO a separate refactoring to do (or at least it should be done for other setting name strings in the manager -- e.g. the string `"stickykeys-enable"` appears 3 times, that would warrant a macro, and it could be shared somehow).

rbuj replied:
Yes, apparently we have a problem with the constants, which are defined in .c files and some of them are not used, so you're right and this can be fixed in another request by moving all constant to one .h file like at-spi2-core does:
```
atspi/atspi-constants.h 
```

rbuj commented:
off-topic: @cwendling can you help us to review https://github.com/mate-desktop/mate-applets/pull/483?

rbuj commented:
@mate-desktop/core-team 

It cannot be merged until the team makes a decision see https://github.com/mate-desktop/mate-settings-daemon/pull/330#discussion_r467255944

sc0w commented:
@rbuj  I think we can wait one week and count the votes

rbuj commented:
@sc0w I don’t know, I’ve waited longer for your reviews. I think the average is about a month.

rbuj commented:
**off-topic**: We should write the decisions taken by the team somewhere, so those members who disagree, they recurrently ask to change the criterion, like using G_GNUC_UNUSED keyword, indents in the code formatting, declaration and assignment at the same line, among others we already talked about.

cwendling commented:
> off-topic: @cwendling can you help us to review [mate-desktop/mate-applets#483](https://github.com/mate-desktop/mate-applets/pull/483)?

I'd be happy to, but I might not have a setup to test this properly before at the very least a week, maybe 3.  I'll try to remember to do it whenever I can.

rbuj commented:
I think the vote can be left open, and it can be implemented later depending on the result, since it's a minor change and there are other requests that depend on this PR.

raveit65 commented:
Hmm, i am too late to party.
Well, i don't really like to vote about a technical issue, a consensus after a discussion might be a better solution.
Honestly, i have no idea if those warnings are really fault positives, another reason not to vote.
IHMO, its a bit weird to use those warnflags to `mate-compiler-flags.m4` from mate-common and disable them in code.

But every rule rule has an exception.
If all you guys thinks that those warnings are fault positives, than it might be OK to disable them in code.
We do the same for Gtk_Action deprecations  warnings in a few packages, because it is impossible to fix them without removing icons from menus before we switch to gtk4.

But i really would like to see that author of PR and `G_GNUC_UNUSED` in code, file out an report at gtk to inform them about fault positive warnings, to clarify the situation.
If not i prefer not to disable them, maybe another developer is able to fix them?

muktupavels commented:
If you want to build with `-Wunused-parameter` you sometimes have only two options, no? Use `G_GNUC_UNUSED` or trick compiler into thinking it is used (`(void) parameter;`)...

For example if you connect to signal that has 5 parameters (p1, p2, p3, p4, p5) and you only need to p1 and p5. You can not remove parameters from your callback function just because you don't need them.

I think it is perfectly fine to mark unused parameters to remove false positive warnings. Otherwise warnings just becomes spam in build log... And you will probably miss real/new warning between false positives.

For example `capslock_beep_callback`. It is connected to `changed` signal. Glib will always call it with 3 parameters - settings, key and user_data. In this example callback is connected specifying key name - `capslock-beep-enable`. So callback will be called only when  `capslock-beep-enable` changes making key parameter uninteresting / unused. It is not glib "bug" because you don't need parameter, it is not code bug because you don't need key in this case. @rbuj, if you say it is not good idea to silence this warning in this case, what is good idea?

rbuj commented:
It depends on the case, as this warning is closely linked to `-Werror=cast-function-type`.

raveit65 commented:
> If you want to build with -Wunused-parameter you sometimes have only two options, no? Use G_GNUC_UNUSED or trick compiler into thinking it is used ((void) parameter;)...
> For example if you connect to signal that has 5 parameters (p1, p2, p3, p4, p5) and you only need to p1 and p5. You can not remove parameters from your callback function just because you don't need them.
>I think it is perfectly fine to mark unused parameters to remove false positive warnings. Otherwise warnings just becomes spam in build log... And you will probably miss real/new warning between false positives.

Ok, than i am fine to mark fault positives as unused parameters in code.
Thx for clarification.

sc0w commented:
I did PR with `G_GNUC_UNUSED` at some points: https://github.com/mate-desktop/mate-settings-daemon/pull/332

raveit65 commented:
> off-topic: We should write the decisions taken by the team somewhere, so those members who disagree, they recurrently ask to change the criterion, like using G_GNUC_UNUSED keyword, indents in the code formatting, declaration and assignment at the same line, among others we already talked about.

Our Wiki?
